### PR TITLE
Add octave and matlab to run module

### LINKED ISSUE #25: Move set matlab log dir to subprocess.run command
the matlab log dir is set in `model.setup()`, see [here](https://github.com/EcoExtreML/STEMMUS_SCOPE_Processing/blob/7f386866550c888b0615d9727c11e8e5e2c86c1c/PyStemmusScope/stemmus_scope.py#L87). It is possible to set it in `subprocess.run()` command using `env` argument. https://docs.python.org/3/library/subprocess.html

### LINKED ISSUE #39: Permission denied in copying input data when running notebooks on WSL
here is the error from when running the [jupyter notebook](https://github.com/EcoExtreML/STEMMUS_SCOPE_Processing/blob/main/notebooks/run_model_in_notebook.ipynb):

```python
---------------------------------------------------------------------------
Error                                     Traceback (most recent call last)
Input In [5], in <cell line: 2>()
      1 # setup the model
----> 2 config_path = model.setup(
      3     ForcingFileName="AR-SLu_2010-2010_FLUXNET2015_Met.nc",
      4     NumberOfTimeSteps="10",
      5 )
      6 print(config_path)

File ~/mamba/envs/pystemmusscope_3.8/lib/python3.8/site-packages/PyStemmusScope/stemmus_scope.py:73, in StemmusScope.setup(self, WorkDir, ForcingFileName, NumberOfTimeSteps)
     70     self._configs["NumberOfTimeSteps"] = NumberOfTimeSteps
     72 # create customized config file and input/output directories for model run
---> 73 _, _, self.cfg_file = config_io.create_io_dir(
     74     self._configs["ForcingFileName"], self._configs
     75     )
     77 # read the run config file
     78 self._configs = config_io.read_config(self.cfg_file)

File ~/mamba/envs/pystemmusscope_3.8/lib/python3.8/site-packages/PyStemmusScope/config_io.py:54, in create_io_dir(forcing_filename, config)
     51 logger.info("%s", message)
     53 # copy model parameters to work directory
---> 54 _copy_data(input_dir, config)
     56 # create output directory
     57 output_dir = work_dir / "output" / f"{station_name}_{timestamp}"

File ~/mamba/envs/pystemmusscope_3.8/lib/python3.8/site-packages/PyStemmusScope/config_io.py:81, in _copy_data(input_dir, config)
     79 for folder in folder_list_vegetation:
     80     os.makedirs(input_dir / folder, exist_ok=True)
---> 81     shutil.copytree(str(config[folder]), str(input_dir / folder), dirs_exist_ok=True)
     83 # copy input_data.xlsx
     84 shutil.copy(str(config["input_data"]), str(input_dir))

File ~/mamba/envs/pystemmusscope_3.8/lib/python3.8/shutil.py:557, in copytree(src, dst, symlinks, ignore, copy_function, ignore_dangling_symlinks, dirs_exist_ok)
    555 with os.scandir(src) as itr:
    556     entries = list(itr)
--> 557 return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
    558                  ignore=ignore, copy_function=copy_function,
    559                  ignore_dangling_symlinks=ignore_dangling_symlinks,
    560                  dirs_exist_ok=dirs_exist_ok)

File ~/mamba/envs/pystemmusscope_3.8/lib/python3.8/shutil.py:513, in _copytree(entries, src, dst, symlinks, ignore, copy_function, ignore_dangling_symlinks, dirs_exist_ok)
    511         errors.append((src, dst, str(why)))
    512 if errors:
--> 513     raise Error(errors)
    514 return dst

Error: [('/mnt/f/P1/US_Var_2001/Vaira_input/directional/', '/mnt/d/test0825/input/AR-SLu_2022-08-31-1412/directional', "[Errno 13] Permission denied: '/mnt/d/test0825/input/AR-SLu_2022-08-31-1412/directional'")]

```

### LINKED ISSUE #44: Make the argument exe_file optional
We use executables with MCR to run the model. However, using octave, this is not needed. We can make the [argument](https://github.com/EcoExtreML/STEMMUS_SCOPE_Processing/blob/d34f8f9c6f6802097be378660124340f2bb81a49/PyStemmusScope/stemmus_scope.py#L37) optional.

### PR DESCRIPTION
**Pull request by SarahAlidoost**
Related to https://github.com/EcoExtreML/STEMMUS_SCOPE/pull/125

- [x] update module
- [x] update docs
- [x] update tests
- [x] update notebooks

Tests:
- [x] crib with octave
- [x] snellius with exe and matlab
- [x] my linux with exe 

closes #44 
closes #39 
closes #25
relates #46

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -19,13 +20,13 @@ Then install the package:
 
 ```sh
 cd STEMMUS_SCOPE_Processing
-pip install -e .
+python3 -m install -e .[dev]
```
Comment (BSchilperoort) [inline CONTRIBUTING.md:23]: This is not the right syntax for Windows users, at least inside a `venv` environment.

If I activate an environment, and check where python and python3 are, this is the result:
```shell
> which python3
/usr/bin/python3

> which python
~/venv/Scripts/python
```
```

**Comment by BSchilperoort** [comment]
I tried running the model using Octave, but it does not seem to work on my Windows machine. Subprocess just returns exit status 1:
```shell
CalledProcessError: Command '['octave --eval "STEMMUS_SCOPE_octave(\'C:\\STEMMUS_SCOPE_data\\workdir\\input\\ZA-Kru_2022-11-03-0927\\ZA-Kru_2022-11-03-0927_config.txt\');exit;"', '--no-gui', '--silent']' returned non-zero exit status 1.
```

**Comment by SarahAlidoost** [comment]
> I tried running the model using Octave, but it does not seem to work on my Windows machine. Subprocess just returns exit status 1:
> 
> ```shell
> CalledProcessError: Command '['octave --eval "STEMMUS_SCOPE_octave(\'C:\\STEMMUS_SCOPE_data\\workdir\\input\\ZA-Kru_2022-11-03-0927\\ZA-Kru_2022-11-03-0927_config.txt\');exit;"', '--no-gui', '--silent']' returned non-zero exit status 1.
> ```

code 1 Indicates that "Action has attempted to execute non-recognized command".  This might be related to how we call octave executable here i.e. `octave` vs `octave.bat` on windows.

**Comment by SarahAlidoost** [comment]
> > I tried running the model using Octave, but it does not seem to work on my Windows machine. Subprocess just returns exit status 1:
> > ```shell
> > CalledProcessError: Command '['octave --eval "STEMMUS_SCOPE_octave(\'C:\\STEMMUS_SCOPE_data\\workdir\\input\\ZA-Kru_2022-11-03-0927\\ZA-Kru_2022-11-03-0927_config.txt\');exit;"', '--no-gui', '--silent']' returned non-zero exit status 1.
> > ```
> 
> code 1 Indicates that "Action has attempted to execute non-recognized command". This might be related to how we call octave executable here i.e. `octave` vs `octave.bat` on windows.

A solution would be to use `shutil.which()` as explained [here](https://docs.python.org/3/library/subprocess.html#popen-constructor).

File: PyStemmusScope/stemmus_scope.py [hunk 2]
```diff
@@ -94,23 +155,32 @@ def run(self) -> str:
...
+            command_line = f'matlab -r "STEMMUS_SCOPE_exe({path_to_config});exit;"'
+            args = [command_line, "-nodisplay", "-nosplash", "-nodesktop"]
+            result = _run_sub_process(args, self.model_src)
+        if self.sub_process=="Octave":
+            # set Octave arguments
+            # use subprocess instead of oct2py,
+            # see issue STEMMUS_SCOPE_Processing/issues/46
+            path_to_config = f"'{self.cfg_file}'"
+            command_line = f'octave --eval "STEMMUS_SCOPE_octave({path_to_config});exit;"'
+            args = [command_line, "--no-gui", "--silent"]
```
Comment (BSchilperoort) [inline PyStemmusScope/stemmus_scope.py:178]:
```suggestion
            eval_code = f'STEMMUS_SCOPE_exe({path_to_config});exit;'
            eval_code = eval_command.replace("\\", "/")
            args = ["octave", "--eval", eval_code, "--no-gui", "--silent"]
```

- Seeing as the Octave package loading stays in `STEMMUS_SCOPE.m` we do not need to use `STEMMUS_SCOPE_octave` here
- `--eval` and its code need to be in separate arguments for Windows
- Octave on Windows requires the double backslashes to be replaced. `.replace` returns the original string if the `\\` are not found

File: CONTRIBUTING.md (hunk 1 above)
Comment (SarahAlidoost) [reply CONTRIBUTING.md:23]: fixed.

File: PyStemmusScope/stemmus_scope.py (hunk 2 above)
Comment (SarahAlidoost) [reply PyStemmusScope/stemmus_scope.py]: @BSchilperoort thanks for fixing the issue. In the updated code, I added `.replace("\\", "/")` to `path_to_config`, hopefully it still works on windows. Can you test it again? thanks.

File: PyStemmusScope/stemmus_scope.py [hunk 3]
```diff
@@ -94,23 +155,34 @@ def run(self) -> str:
...
+            result = _run_sub_process(args, self.model_src)
+        if self.sub_process=="Octave":
+            # set Octave arguments
+            # use subprocess instead of oct2py,
+            # see issue STEMMUS_SCOPE_Processing/issues/46
+            path_to_config = f"'{self.cfg_file}'"
+            # fix for windows
+            path_to_config = path_to_config.replace("\\", "/")
+            command_line = f'octave --eval "STEMMUS_SCOPE_exe({path_to_config});exit;"'
+            args = [command_line, "--no-gui", "--silent"]
```
Comment (BSchilperoort) [inline PyStemmusScope/stemmus_scope.py:180]:
```suggestion
            eval_code = f'STEMMUS_SCOPE_exe({path_to_config});exit;'
            args = ["octave", "--eval", eval_code, "--no-gui", "--silent"]
```

This still has to be changed for it to work on Windows. All the arguments have to be passed separately to subprocess for it to work.

File: PyStemmusScope/stemmus_scope.py [hunk 4]
```diff
@@ -94,23 +155,34 @@ def run(self) -> str:
...
+            # run using MCR
+            args = [f"{self.exe_file} {self.cfg_file}"]
+            # set matlab log dir
+            os.environ['MATLAB_LOG_DIR'] = str(self._config["InputPath"])
+            result = _run_sub_process(args, None)
+        if self.sub_process=="Matlab":
+            # set Matlab arguments
+            path_to_config = f"'{self.cfg_file}'"
+            command_line = f'matlab -r "STEMMUS_SCOPE_exe({path_to_config});exit;"'
+            args = [command_line, "-nodisplay", "-nosplash", "-nodesktop"]
```
Comment (BSchilperoort) [inline PyStemmusScope/stemmus_scope.py:170]:
```suggestion
            eval_code= f"STEMMUS_SCOPE_exe({path_to_config});exit;"
            args = ["matlab", "-r", eval_code, "-nodisplay", "-nosplash", "-nodesktop"]
```
This change is probably also required for the Matlab, for Windows compatibility. See the suggestion on lines 179-180.

File: PyStemmusScope/stemmus_scope.py [hunk 5]
```diff
@@ -12,32 +13,95 @@
...
+        sub_process(str, optional): use `Matlab` or `Octave`. It is optional if
+        `model_src_path` is a path to Stemmus_Scope executable file.
 
     Example:
         See notebooks/run_model_in_notebook.ipynb in `STEMMUS_SCOPE_Processing
         repository <https://github.com/EcoExtreML/STEMMUS_SCOPE_Processing>`_
     """
 
-    def __init__(self, config_file: str, exe_file: str):
+    def __init__(self, config_file: str, model_src_path: str, sub_process: str = None):
```
Comment (BSchilperoort) [inline PyStemmusScope/stemmus_scope.py:88]: I'm not sure if `sub_process` is the right name here. I would think [`interpreter`](https://en.wikipedia.org/wiki/Interpreter_(computing)) is more accurate.

File: PyStemmusScope/stemmus_scope.py [hunk 6]
```diff
@@ -12,32 +13,95 @@
...
+
 
 class StemmusScope():
     """PyStemmusScope wrapper around Stemmus_Scope model.
     see https://gmd.copernicus.org/articles/14/1379/2021/
 
-    It sets the model with a configuration file and executable file.
-    It also prepares forcing and soil data for model run.
+    It sets the model with a configuration file and executable file. It also
+    prepares forcing and soil data for model run.
```
Comment (BSchilperoort) [inline PyStemmusScope/stemmus_scope.py:72]:
```suggestion
   Configures the model and prepares forcing and soil data for the model run.
```

File: PyStemmusScope/stemmus_scope.py [hunk 7]
```diff
@@ -12,32 +13,95 @@
 
 logger = logging.getLogger(__name__)
 
+def _is_model_src_exe(model_src_path: Path):
+    #TODO add docstring
```
Comment (BSchilperoort) [inline PyStemmusScope/stemmus_scope.py:17]: Do you still want to add the docstring for this PR?

File: PyStemmusScope/stemmus_scope.py [hunk 8]
```diff
@@ -12,32 +13,95 @@
...
 
     Args:
         config_file(str): path to Stemmus_Scope configuration file. An example
         config_file can be found in tests/test_data in `STEMMUS_SCOPE_Processing
         repository <https://github.com/EcoExtreML/STEMMUS_SCOPE_Processing>`_
-        exe_file(str): path to Stemmus_Scope executable file.
+        model_src_path(str): path to Stemmus_Scope executable file or to a
+        directory containing model source codes.
+        sub_process(str, optional): use `Matlab` or `Octave`. It is optional if
+        `model_src_path` is a path to Stemmus_Scope executable file.
```
Comment (BSchilperoort) [inline PyStemmusScope/stemmus_scope.py:81]:
```suggestion
        sub_process(str, optional): use `Matlab` or `Octave`. Only required if
        `model_src_path` is a path to Stemmus_Scope executable file.
```

**Review by BSchilperoort — CHANGES_REQUESTED** [review]
Hi Sarah, you missed part of my suggestion which was to split up the `octave --eval CODE` into separate `octave`, `--eval`, and `CODE` list elements. This is required for it to work on Windows.

I also reviewed the rest of the PR and left some suggestions and comments! 👍

File: PyStemmusScope/stemmus_scope.py (hunk 3 above)
Comment (SarahAlidoost) [reply PyStemmusScope/stemmus_scope.py:180]: ah sorry. I commit your suggestion here, but I will check if the code works on linux too.

File: PyStemmusScope/stemmus_scope.py (hunk 4 above)
Comment (SarahAlidoost) [reply PyStemmusScope/stemmus_scope.py:170]: I commit your suggestion here, but I will check if the code works on linux too.

File: PyStemmusScope/stemmus_scope.py (hunk 8 above)
Comment (SarahAlidoost) [reply PyStemmusScope/stemmus_scope.py]: Fixed!

File: PyStemmusScope/stemmus_scope.py (hunk 7 above)
Comment (SarahAlidoost) [reply PyStemmusScope/stemmus_scope.py]: Added.

File: PyStemmusScope/stemmus_scope.py (hunk 5 above)
Comment (SarahAlidoost) [reply PyStemmusScope/stemmus_scope.py]: I like the suggestion. Replaced.

File: PyStemmusScope/stemmus_scope.py (hunk 4 above)
Comment (SarahAlidoost) [reply PyStemmusScope/stemmus_scope.py]: As discussed the separate args don't work on Linux. I used `shlex.join()` for it.

**Review by BSchilperoort — APPROVED** [review]
Hi Sarah, great work! Octave + Windows 10 works as intended now 😄.

~~Seeing as you'll be gone for a week, and to avoid merge conflicts, I will merge this PR once I fixed the Windows CI.~~ The tests are failing due to a lack of Windows wheels for the netcdf 1.6.0 package. I will fix it to <1.6.0.

**Comment by BSchilperoort** [comment]
I noticed that the tests were failing on Windows, as trying to use a non-existent EXE results in a `FileNotFoundError` on Windows, instead of the `ValueError` you expected. I have updated the tests to account for this.

See the changes here:
https://github.com/EcoExtreML/STEMMUS_SCOPE_Processing/pull/45/commits/6a0d02480c1659acafbd3fb0490805318191d9b7#diff-d57ca3aa8acb1482e37888eb52a0746872c9c36e6bc8c263dd3a1efed5f3196fR14-R23

**Comment by SarahAlidoost** [comment]
> I noticed that the tests were failing on Windows, as trying to use a non-existent EXE results in a `FileNotFoundError` on Windows, instead of the `ValueError` you expected. I have updated the tests to account for this.
> 
> See the changes here: [6a0d024#diff-d57ca3aa8acb1482e37888eb52a0746872c9c36e6bc8c263dd3a1efed5f3196f](https://github.com/EcoExtreML/STEMMUS_SCOPE_Processing/commit/6a0d02480c1659acafbd3fb0490805318191d9b7#diff-d57ca3aa8acb1482e37888eb52a0746872c9c36e6bc8c263dd3a1efed5f3196f)

Thanks for the review and fixes. I will merge the PR.

**Pull request merged**